packages/engine/src/engine.cairo

```diff
@@ -203,6 +205,43 @@ pub impl EngineExtrasImpl<T, +Drop<T>> of EngineExtrasTrait<T> {
         };
         return sub_script;
     }
+
+    fn check_error_condition(ref self: Engine<T>, final_script: bool) -> Result<(), felt252> {
+        // Check if execution is actually done
+        if self.script_idx < self.scripts.len() {
+            return Result::Err(Error::SCRIPT_UNFINISHED);
+        }
+
+        // Check if witness stack is clean
+        if final_script
+            && self.is_witness_active(0)
+            && self.dstack.len() != 1 { // TODO: Hardcoded 0
+            return Result::Err(Error::SCRIPT_NON_CLEAN_STACK);
+        }
+        if final_script
+            && self.has_flag(ScriptFlags::ScriptVerifyCleanStack)
+            && self.dstack.len() != 1 {
+            return Result::Err(Error::SCRIPT_NON_CLEAN_STACK);
+        }
+
+        // Check if stack has at least one item
+        if self.dstack.len() < 1 {
+            return Result::Err(Error::SCRIPT_EMPTY_STACK);
+        }
+
+        // Check the final stack value
+        match self.dstack.pop_bool() {
```

Comment:
I think this pop_book instead of peeking the top stack element might be causing a lot of the test failures.